Interactive view

huyen-nguyen opened the pull request.

huyen-nguyen commented:
https://github.com/user-attachments/assets/7760cae7-678e-49f2-adcc-7fb36a7a2f70

huyen-nguyen commented:
https://github.com/user-attachments/assets/b1a5a5c9-93c2-424e-b3a8-20317384a5da

thomcsmits reviewed (changes requested):
Great addition! :) I resolved the issues with main. Note that everything for the modal view is now separated into `ExampleDetailView.jsx`. 

Did you look at the sizing of the images in the gallerypanel? It seems like it's off:

https://github.com/user-attachments/assets/e59a4af6-5fcf-4082-871a-1ee35debaa4a

huyen-nguyen commented:
Thanks for setting the modal view aside, @thomcsmits!

The sizing of image is meant to be that way, I believe. The width of image/text/spec panels are the same regardless of scaling.

thomcsmits commented:
> The sizing of image is meant to be that way, I believe. The width of image/text/spec panels are the same regardless of scaling.

Wouldn't it be better if the text & spec boxes would be the same height as the image box?

![image](https://github.com/user-attachments/assets/d4d768e6-6263-4c33-86bb-c06d9d14c31d)

huyen-nguyen commented:
@sehilyi In the latest commit above (https://github.com/huyen-nguyen/geranium/pull/36/commits/913ed504d35fc397e1a0069b240a066070d33e43), the editor automatically formats JSON. 

I used debounce to wait until the typing stops for 800ms before trying to parse and then format/beautify it. We might use a similar technique to help formatting in the Gosling editor.



https://github.com/user-attachments/assets/07c04538-579f-44d5-8993-5cc29fca4f57

huyen-nguyen commented:
Added live Gosling editor, we can now directly edit any examples in this system: from landing page, search results or gallery. By clicking on an example, it will open a modal containing Gosling editor, resulting visualization, and initial text description.

Automatically format any JSON specifications whether in query or result. Minimap on the right of spec editor and error indicator when spec query is not a valid JSON object (thanks to Monaco).

Other updates include: 
- 3 themes installed. 
- Movable divider between editor and visualization. 
- In gallery/landing page: added shadow upon hovering instead of a repeated expand icon for every images. 
- Closing modal via an `x` button instead of clicking outside of modal area -- to prevent user from accidentally closing editor while authoring.

huyen-nguyen commented:
Hey @sehilyi and @thomcsmits, could you take a look at the latest update in this PR? There are quite a few changes, so please make sure the npm packages are updated. I ended up merging it to main without much thought, but I'm open to reopening it if any issues come up. 

Thanks a lot!
Huyen
